atomic_reactor/plugins/pre_change_from_in_df.py

```diff
@@ -29,21 +39,54 @@ def __init__(self, tasker, workflow):
         super(ChangeFromPlugin, self).__init__(tasker, workflow)
 
     def run(self):
-        """
-        run the plugin
-        """
-        base_image = self.workflow.builder.base_image
-        try:
-            base_image_id = self.workflow.base_image_inspect['Id']
-        except KeyError:
-            self.log.error("Id is missing in inspection: '%s'", self.workflow.base_image_inspect)
-            raise
-        self.log.debug("using base image '%s', id '%s'", base_image, base_image_id)
-        for line in fileinput.input(self.workflow.builder.df_path, inplace=1):
-            re_match = re.match(r"^FROM .+$", line)
-            if re_match:
-                new_from = "FROM %s" % base_image_id
-                sys.stdout.write(new_from + '\n')
-                self.log.info("changed FROM: '%s' -> '%s'", re_match.group(0), new_from)
-            else:
-                sys.stdout.write(line)
+        builder = self.workflow.builder
+        dfp = DockerfileParser(builder.df_path)
+
+        df_base = dfp.baseimage
+        build_base = builder.base_image.to_str()
+        if build_base != df_base and build_base != builder.parent_images[df_base]:
+            # something updated the build's base_image without updating the corresponding
+            # parent_images entry; that should be OK, let base_image override.
+            builder.parent_images[df_base] = build_base
```

Comment:
Should `builder.base_image` be a property which handles updating `self.parent_images`?

Reply:
I debated that a few times and ultimately decided not to. However I'm reconsidering now because it rested on not having ready access everywhere to know which key in `parent_images` was from the Dockerfile baseimage. Since later I made `set_base_image` also set `original_base_image` with that info, that now seems like a lame excuse for not keeping the two in sync automatically.